### ISSUE TITLE
Builtin for getting real time in nanoseconds

### ISSUE THREAD

**Issue by viktormalik** [iovisor/bpftrace#2512]
### Is your feature request related to a problem? Please describe.

As was suggested in discussion #2486, there may be applications that require bpftrace programs to get the current time in nanoseconds. Since kernel 6.2, there is the [bpf_ktime_get_tai_ns helper](https://elixir.bootlin.com/linux/v6.2/source/kernel/bpf/helpers.c#L204) that allows BPF programs to access the TAI clock.

### Describe the solution you'd like

A new builtin variable, e.g. `nsecs_tai` or simply `tai` that would call the `bpf_ktime_get_tai_ns` helper (if it is available on the current kernel) and return the TAI in nanoseconds. Then, `strftime` can be used to create a formatted timestamp.

**Comment by chengshuyi** [comment]
Hi, i'm learning the bpftrace code, I can add this function if you don't mind.

**Comment by viktormalik** [comment]
Please do! Thanks :-)

**Comment by ajor** [comment]
With this new TAI clock helper, is there any reason anyone would want to use the old nsecs-since-boot timestamp? Can we just switch the "nsecs" builtin across to using this new clock?

**Comment by chengshuyi** [comment]
The kernel currently provides four methods for obtaining timestamps, namely bpf_ktime_get_ns, bpf_ktime_get_boot_ns, bpf_ktime_get_coarse_ns and bpf_ktime_get_tai_ns. They can be roughly divided into two categories:
1. bpf_ktime_get_ns, bpf_ktime_get_boot_ns and bpf_ktime_get_coarse_ns are  the time elapsed since system boot》
2. bpf_ktime_get_tai_ns is to get the time of CLOCK_TAI.

So, I think it may be necessary to provide methods to get these two different types of timestamps.

Thanks!

**Comment by ajor** [comment]
What would the use cases be in bpftrace for exposing both these categories of timestamps?

**Comment by chengshuyi** [comment]
I don't have a specific use case, and I don't see any special requirements for the time elapsed since system boot from the issues and pr of bpftrace. On the contrary, #1471  and #2486 have demand for tai time. Among them, #1471 uses the "triple vdso sandwich" method to obtain approximate tai time.

**Comment by viktormalik** [comment]
> With this new TAI clock helper, is there any reason anyone would want to use the old nsecs-since-boot timestamp? Can we just switch the "nsecs" builtin across to using this new clock?
> 
> Although, being TAI it's currently 37 seconds off UTC and will only deviate more... this might end up being confusing. I can't see any helper which returns UTC, so using TAI by default still might be the best option for now.

The only problem I can see is if someone is using `nsecs` to get the absolute time from boot, not just for calculating delta. I can't think of any use-case where that would be helpful but if someone is doing it, we will break their script by changing the semantics of `nsecs`. Still probably worth doing.

**Comment by chengshuyi** [comment]
@viktormalik Hi, `strftime([fmt], nsec)` now returns tai time via #1471 which uses the "triple vdso sandwich" method to obtain approximate tai time. So, now strftime(fmt, nsec_tai) will print abnormal values.

```c
BEGIN { 
    $boot_ts = strftime("%m/%d/%y %H:%M:%S", nsecs); 
    $tai_ts = strftime("%m/%d/%y %H:%M:%S", nsecs_tai); 
    
    printf("boot time: %s\n tai time: %s\n", $boot_ts, $tai_ts);
}
```

Output is :

```shell
boot time: 05/10/23 16:27:22
 tai time: 09/04/76 02:27:35
```

My question is whether it is possible to change the meaning of strftime, so that it does not add boottime_.

**Comment by viktormalik** [comment]
Resolved by #2619.
